Basic sync support

grigoryk opened the pull request:
Fast-forward syncing with support for "merging" the bootstrap transaction.

Issues more or less resolved with this: #513, #514, #511

rnewman commented on src/conn.rs:
```diff
@@ -451,9 +458,56 @@ impl Queryable for Store {
+        let mut last_tx_uuid = None;
+        // Giant hack: we're expecting 'parts' to not actually contain metadata
+        // about any transactions - just their datoms! we generate metadata ourselves
+        // while transacting, thus loosing when transactions were made originally /o\
```
Yep, for that you'll need to go a little lower, and also renumber (#494).

rnewman commented on src/conn.rs:
```diff
@@ -451,9 +458,56 @@ impl Queryable for Store {
+        match tx_list {
+            Some(list) => {
+                for tx in list {
+                    let in_progress = self.begin_transaction()?;
```
I think you should lift this out of the `for` loop — you can transact multiple times against an `in_progress`, and they'll all share a single SQLite transaction.

rnewman commented on src/conn.rs:
```diff
@@ -451,9 +458,56 @@ impl Queryable for Store {
+        let mut tx_list = None;
+        {
+            let mut db_tx = self.sqlite.transaction()?;
+            if let Some(list) = Syncer::flow(&mut db_tx, server_uri, &uuid)? {
```
If you implement `flow` on `InProgress`, you get the transaction for free… the only tricky bit is that you'd ideally like to not do any work (including establishing an exclusive transaction!) if there's no data to write.

rnewman commented on src/conn.rs:
```diff
@@ -451,9 +458,56 @@ impl Queryable for Store {
+            db_tx.commit()?;
+        }
+        let mut last_tx_entid = None;
+        let mut last_tx_uuid = None;
```
Use a pair to indicate that these two things must be set together.

```
let mut last_tx = None;
…
last_tx = Some((uuid, entid));
```

rnewman commented on tolstoy/src/syncer.rs:
```diff
@@ -169,6 +171,13 @@ impl<'c> TxReceiver for UploadingTxReceiver<'c> {
+#[derive(Debug)]
+pub struct Tx {
+    pub tx: Uuid,
+    pub parts: Vec<TxPart>
```
Nit: trailing comma, please.

rnewman reviewed (commented):
Keep going!

rnewman commented on src/conn.rs:
```diff
@@ -157,6 +154,12 @@ impl Store {
         })
     }
+
+    pub fn fast_forward_user_partition(&mut self, new_head: Entid) -> Result<()> {
```
What I expect to see here is essentially a compare-and-set: if my local partition map is _this_, then expand it to _this one_.

Doing that work requires writing to the DB, so it shouldn't be on `Store`, it should be on `InProgress`; if the DB write fails the metadata will be discarded.

And you want to advance all parts — typically `user` and `tx` — at the same time, atomically, right? Remember that the data itself will refer to tx IDs — it's not enough to only work with the user partition!

rnewman commented on src/conn.rs:
```diff
@@ -157,6 +154,12 @@ impl Store {
+    pub fn fast_forward_user_partition(&mut self, new_head: Entid) -> Result<()> {
+        let mut metadata = self.conn.metadata.lock().unwrap();
+        metadata.partition_map.expand_up_to(":db.part/user", new_head);
+        db::update_partition_map(&mut self.sqlite, &metadata.partition_map).map_err(|e| e.into())
```
I would rather see the metadata change as a side-effect of updating the DB. The `db::` code is where we'll be able to do compare-and-set, and we already have code to update the metadata after a `transact`.

rnewman replied:
Relatedly: it might be worth thinking about what you're doing here as a special case of renumbering.

Renumbering is the process of taking a space of new remote identifiers, moving any conflicting local identifiers out of the way, and making space locally for the remote identifiers. If there are no conflicting local identifiers, then the second step involves no work.

This is a nice unification because it reduces the number of different code paths…

rnewman commented on src/lib.rs:
```diff
@@ -95,6 +97,13 @@ pub mod conn;
+#[cfg(feature = "syncable")]
+pub mod sync;
+
+pub fn get_name() -> String {
```
I'm  pretty sure I deleted this… it was just there for testing, long ago.

rnewman commented on src/sync.rs:
```diff
@@ -0,0 +1,133 @@
+// Copyright 2016 Mozilla
```
2018

rnewman commented on src/sync.rs:
```diff
@@ -0,0 +1,133 @@
+        // partition's next allocation will not overlap with just-inserted datoms.
+        // To allow for "holes" in the user partition (due to data excision),
+        // we track the highest incoming entid we saw, and expand our
+        // local partition to match.
```
Again, I think this process would benefit from splitting out renumbering.

1. Download the new remote parts table.
2. Renumber locally, which also makes space locally for remotely allocated identifiers. (In the fast-forward case no work is done here.)

Now we get into the state machine. For fast-forwarding:

3. Apply remote datoms (which now don't exceed the parts table!).

Make sense?

ncalexan commented on src/errors.rs:
```diff
@@ -15,6 +15,7 @@ use std; // To refer to std::result::Result.
 use std::collections::BTreeSet;
 
 use rusqlite;
+use uuid;
```
Lift this commit out and land it, since it's mechanical.

ncalexan commented on tolstoy/src/metadata.rs:
```diff
@@ -41,6 +46,14 @@ impl SyncMetadataClient {
         Ok(())
     }
+
+    pub fn get_partitions(tx: &rusqlite::Transaction) -> Result<PartitionMap> {
```
We have https://github.com/mozilla/mentat/blob/5fe4f12d32379415e36eb80877cbeec6fd443ffa/db/src/db.rs#L443 just waiting to be generalized; let's use it.

ncalexan commented on tolstoy/src/schema.rs:
```diff
@@ -29,6 +37,12 @@ pub fn ensure_current_version(tx: &mut rusqlite::Transaction) -> Result<()> {
+    // For now, only information about the user partition is tracked.
+    // We rely on the transactor to advance the tx partition.
+    // Support for migrating the db partition forward (that is, bootstrap schema) is TBD.
+    tx.execute("INSERT OR IGNORE INTO tolstoy_parts VALUES (?, ?, ?)", &[&PARTITION_USER, &USER0, &USER0])?;
```
We also have https://github.com/mozilla/mentat/blob/5fe4f12d32379415e36eb80877cbeec6fd443ffa/db/src/bootstrap.rs#L88.

ncalexan commented on tolstoy/src/schema.rs:
```diff
@@ -98,6 +122,12 @@ pub mod tests {
             }
         }
 
+        let new_idx = USER0 + 1;
```
There's a function for writing the partition map in general, too.

ncalexan commented on tolstoy/src/schema.rs:
```diff
@@ -80,6 +96,14 @@ pub mod tests {
             (_, _) => { panic!("Wrong number of results."); },
         }
+
+        let partitions = SyncMetadataClient::get_partitions(&tx).unwrap();
```
In general, we can't be `:db.part/user`-centric, 'cuz `:db.part/tx` advances too.  I think this will get simpler as you generalize, not more complicated.

ncalexan commented on tolstoy/src/syncer.rs:
```diff
@@ -143,8 +147,27 @@ impl<'c> TxReceiver for UploadingTxReceiver<'c> {
+        }
+
+        // Annotate first datom in the series with the user partition information.
+        // TODO this is obviously wrong - we want to read partition info without
```
It feels like this is at the wrong place in the stack.  You want to change the flow from `Vec<Tx>` to `(PartitionMap, Vec<Tx>)`, but what you've done is change it to `Vec<(Option<PartitionMap>, Tx)>` and an assumption about option.

It feels to me like we need some notion of a *chunk* that's more general than what we have now; some chunks are transaction data, yes; some are transactions themselves; but some are also partition maps.

I think you're probably aware of this, and tomorrow we can talk through the difficulty of changing the underlying model more broadly.

grigoryk replied:
I'm viewing this particular change as purely temporary - I really don't like having partition information one step removed from the transaction - and following the path of least resistance for now. The goal is to move this out of chunks entirely, and lift it up into transactions. They should be annotated with partitionmaps, and the only reason that's not the way it is now is lack of server support.

Currently, I don't think there's any _other_ need to generalize a chunk, and doing so seems like a distraction.

My higher level view of this - let's settle on the overall flow first - how data generally moves around, what do we need to move around, etc - then re-shape the server and clients to match. In the meantime, iterate quickly and maintain some kind of momentum.

"Nothing is as permanent as the temporary", but I hope that won't apply :dancing_men:

ncalexan commented on db/src/renumber.rs:
```diff
@@ -300,10 +300,16 @@ mod tests {
 
         // N.b., after renumbering, the on-disk data is not reflected in the `TestConn`!
-        let mut db = renumber(&local.sqlite,
+        let mut db;
```
I think something has gotten confused.  Everything here is going to take place inside an encapsulating SQL transaction; we shouldn't be managing transactions in this helper function at all.  It's up to consumers to manage isolation, etc; and `sync` (like `transact`) will do so.

Reason to yourself: suppose renumbering succeeds but something later fails -- how will I roll back renumbering?  If you commit, you can't.  (Without using nested transactions or checkpoints or other complicated things, which I really don't think we need.)

ncalexan reviewed (commented):
We'll talk more tomorrow, @grigoryk.  Thanks for pushing the spear on this!

grigoryk commented:
Rebased on top of master & the timelines work in #783, folded history together as it's not really valuable and needs to be retold.

ncalexan reviewed (approved):
I see what you've done -- you've inserted the Tolstoy create in between the top-level crate and all the other crates.

But there's another approach: push the details of Mentat transactions (most of `Conn`) down closer to the `db` crate.  That is, the `db` crate basically "is" the `transaction` crate  right now; we just do a bunch of work gluing together bits of `db` at top-level right now.

I don't know what the right approach is.  My mental model was that the `db` crate would end up owning more of the metadata-handling functionality over time.  But I'm not really against this approach.  Have you thought through both?  Are you confident that this split is the one we should pursue?

One thing that's not clear to me is that there will ever be a _second_ Tolstoy like feature built on Mentat.  That is, a second feature that is _inside_ Mentat but wants the high-level query and transaction interfaces.  But as I type that, I realize I'm being short-sighted: **of course** there's going to be a feature like that, I just don't know what it is yet.

Give me your perspective, and then roll on.  (I've reviewed up to the "split" Pre:.)

grigoryk commented:
On the `db` crate - yeah, I briefly thought about it and I decided that I don't have a good enough sense yet for how to optimally organize this stuff, but splitting high level metadata away from the top level _is_ needed in some form.

My main goal was to separate that stuff out of the top-level crate, to enable nicer patterns for the internal consumers (well, just tolstoy for now). If we decide that it's better to keep shuffling this stuff a bit more to find it a better home, at least it _should_ be easier from this point on, and the consumers building on top of the `transaction` crate won't be affected significantly - likely just a few import changes.

You're right that what's in the `transaction` right now is a thin layer on top of `db`, doing some bookkeeping one layer above it. I'm also not a fan of having `Store` and `Conn` at the top level, either - it makes writing tests that want to use a real `Conn` awkward, forcing them to be at the top level, away from the functionality that's being tested. The `TestConn` surrogate is a smell of that. So, there's more re-shuffling to be done! My hope is that little steps like this will inform what the ideal world looks like.

grigoryk commented on db/src/lib.rs:
```diff
@@ -48,6 +48,7 @@ mod bootstrap;
 pub mod entids;
 pub mod internal_types;    // pub because we need them for building entities programmatically.
 mod metadata;
+mod renumber;
```
I don't actually use it currently, no. Or rather, it's _used_ in a trivial case of fast-forwarding local as a way of expanding a local partition, but that operation is a no-op with the "derived partition maps" changes.

Commit:
Pre: remove remnants of 'open_empty'

Commit:
Part 2: Support basic "side-effects" syncing

This patch introduces a concept of a follow-up sync. If a sync generated
a "merge transaction" (a regular transaction that contains assertions
necessary for local and remote transaction logs to converge), then
this transaction needs to be uploaded in a follow-up sync.

Generated SyncReport indicates if a follow-up sync is required.

Follow-up sync itself is just a regular sync. If remote state did not change,
it will result in a simple RemoteFastForward. Otherwise, we'll continue
merging and requesting a follow-up.

Schema alterations are explicitly not supported.

As local transactions are rebased on top of remote, following changes happen:
- entids are changed into tempids, letting transactor upsert :db/unique values
- entids for retractions are changed into lookup-refs if we're confident they'll succeed
-- otherwise, retractions are dropped on the floor

Commit:
Pre: Cleanup 'datoms' table after a timeline move

Since timeline move operations use a transactor, they generate a
"phantom" 'tx' and a 'txInstant' assertion. It is "phantom" in a sense
that it was never present in the 'transactions' table, and is entirely
synthetic as far as our database is concerned.
It's an implementational artifact, and we were not cleaning it up.

It becomes a problem when we start inserting transactions after a move.
Once the transactor clashes with the phantom 'tx', it will retract the
phantom 'txInstant' value, leaving the transactions log in an incorrect state.

This patch adds a test for this scenario and elects the easy way out: simply
remove the offending 'txInstant' datom.

Commit:
Part 1: Sync without support for side-effects

A "side-effect" is defined here as a mutation of a remote state as part
of the sync.

If, during a sync we determine that a remote state needs to be changed, bail out.

This generally supports different variations of "baton-passing" syncing, where clients
will succeed syncing if each change is non-conflicting.

grigoryk commented:
Wow, this has been open for 7 months. This is a good example of how "not to PR". Anyway...

The current state is as follows.

Part 1 is very non-controversial. It clears up the code base, introduces bunch of internal types and solidifies the general flow of things. As for mechanics of a sync beyond fast-forward scenarios, only a "no-op rebase of local on top of remote" is supported. That is, if a merge generates any assertions/retractions, we bail out.

A merge is essentially: "rewind local transactions, transact the remote transactions, then transact local. succeed if local transactions had no effect, bail out otherwise".

Part 2 takes things further, and generally allows anything one can throw at a transactor (sans implemention bugs and some TODOs). Except for schema alterations. It's fine to add new vocabulary (including at the same time, on both "remote" and "local", as long as it's defined the same way), but we bail out if we see changes to existing vocabulary. This is done largely so that we can address syncing of schema alterations by themselves.

A merge is essentially: "rewind local transactions, transact the remote transactions, then ensure local may be transacted, perform some modifications, transact them, and request a follow-up sync if that operation produced any new datoms."

Certain choices were made around handling of entids and retractions, which are non-optimal in some cases but produce understandable results - and are natural given the current approach.

General approach is to let the transactor handle merges, letting the schema dictate how that happens. A concrete manifestation of this:
- :db/unique attributes will upsert. Asserting entities against non-unique attributes will naturally produce duplicates.

During a rebase, local entids are replaced with tempids. Retractions pose a challenge: a retraction won't allocate, so if a tempid mentioned in the retraction doesn't resolve, transaction will fail. If a retraction datom is by itself, its tempids will never resolve. Current approach is to replace tempids with lookup-refs whenever we see retractions that use tempids that we recognize won't resolve. Additionally, a lookup-ref mentioned in a retraction must resolve, and so we drop on the floor any retractions that fail to meet that criteria.

While all of this has certain limitations, it has very nice general property: an intuition about transactor's behaviour applies directly to sync's behaviour. Additionally, expanding what transactor accepts and supports will expand sync's capabilities with little additional effort. For example, if we ever add compound uniqueness, above sync mechanics should "just work" as you'd expect them.

grigoryk commented:
Currently tests are failing due to a bug in the transactor: https://github.com/mozilla/mentat/issues/818

Commit:
Post: use a macro for more readable tests

Commit:
Tolstoy README

ncalexan commented on tolstoy/README.md:
nit: "[must be the same] on all participating clients (i.e., core schema alterations are unsupported)."

ncalexan commented on tolstoy/README.md:
nit: this sentence got caught in an edit.

grigoryk replied:
Oh man, it did indeed.

ncalexan commented on tolstoy/README.md:
nit: "[of] the [shared root]".

ncalexan commented on tolstoy/README.md:
nit: "The [first three] ...".

ncalexan commented on tolstoy/README.md:
nit: "[about] the [transactor's]"

ncalexan commented on tolstoy/README.md:
s/additions/assertions/

ncalexan commented on tolstoy/README.md:
`:db/unique` (include colon)

ncalexan commented on tolstoy/README.md:
by the remote client

ncalexan commented on tolstoy/README.md:
nit: `:db/ident` (throughout -- keep the colons).

ncalexan commented on tolstoy/README.md:
"in a basic manner"

ncalexan commented on tolstoy/README.md:
Can you include a hash, so that we know where this left off?

ncalexan reviewed (approved):
Looking good.  Some small changes, then bombs away!
